### PR TITLE
Miri: improve spans of required_const failures

### PR DESCRIPTION
**Pull request by RalfJung**
In https://github.com/rust-lang/rust/pull/75339 I added a loop evaluating all consts required by a function body. Unfortunately, if one of their evaluations fails, then the span used for that was that of the first statement in the function body, which happened to work form some existing test but is not sensible in general.

This PR changes it to point to the whole function instead, which is at least not wrong.

r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
To do better, we'd have to either
* pass a span through the error machinery (i.e., add a `Span` field to `InterpError`) which (if present) overwrites the span of the topmost frame, or
* make it so that the frame knows it is currently evaluating a const, and can report its span.

In the light of [this chaos](https://github.com/rust-lang/rust/issues/71800#issuecomment-671805859), I am not sure I want to add yet another `Span` field to errors, and such "overwriting" behavior is often confusing. So if we think that we should try to do better, my proposal would be to replace this field of `Frame`
```rust
    pub loc: Option<mir::Location>,
```
by something of a type like `Result<mir::Location, Span>`, so that we can say what the span is when we are not currently evaluating a statement in that frame.

**Comment by oli-obk** [comment]
Yes, I also prefer that proposal. Although `Result` isn't the best type here, I'm not sure the situation warrants its own enum (though that could make it more obvious that something unusual is going on here). The current `None` situation could be represented by `Err(DUMMY_SP)`

**Comment by RalfJung** [comment]
Or rather, `Err(self.body.span)`. Yes.

**Comment by RalfJung** [comment]
All right, I implemented that now.

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me with nits

File: src/test/ui/consts/const-err-multi.stderr [hunk 1]
```diff
@@ -24,17 +24,17 @@ error: any use of this value will cause an error
   --> $DIR/const-err-multi.rs:7:19
    |
 LL | pub const C: u8 = A as u8;
-   | ------------------^^^^^^^-
+   | ------------------^-------
```
Comment (oli-obk) [inline src/test/ui/consts/const-err-multi.stderr:27]: this is great!

File: src/librustc_mir/interpret/intrinsics/caller_location.rs [hunk 2]
```diff
@@ -30,8 +30,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             // Assert that there is always such a frame.
             .unwrap();
         // Assert that the frame we look at is actually executing code currently
-        // (`current_source_info` is None when we are unwinding and the frame does
-        // not require cleanup).
+        // (`loc` is None when we are unwinding and the frame does not require cleanup).
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/intrinsics/caller_location.rs:33]: (edited to `Err`)

**Comment by RalfJung** [comment]
@bors r-
investigating branch CI failure

**Comment by RalfJung** [comment]
Hu, a test failure but local runs don't show anything??

**Comment by RalfJung** [comment]
Ah looks like that test was changed recently even though it already existed before... very strange.

**Comment by RalfJung** [comment]
@oli-obk for some reason `uninhabited-const-issue-61744.stderr` changed a lot. No idea what is happening.

**Comment by RalfJung** [comment]
Also how horrible is GitHub? Why can I not go to the git log from https://github.com/rust-lang/rust/blob/d45dcb509527d847c7c3bf459a0e796801067b33/src/test/ui/consts/uninhabited-const-issue-61744.stderr ? I thought that used to be possible?

**Comment by RalfJung** [comment]
Hm, that test also did not change recently. So it must be some other change on the rustc side that makes it such that in combination with this PR, the output for that test changes. WTF.

**Comment by RalfJung** [comment]
Oh I see, this is a stack overflow, so it likely interacts with https://github.com/rust-lang/rust/pull/75338.

File: src/librustc_mir/interpret/eval_context.rs [hunk 3]
```diff
@@ -683,8 +694,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         }
         // done
         self.frame_mut().locals = locals;
-
         M::after_stack_push(self)?;
+        self.frame_mut().loc = Ok(mir::Location::START);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs:698]: It's about the order of these two lines: for errors that happen during `after_stack_push`, so we want to use the span of the first statement, or the span of the entire body? IMO the latter makes more sense, the first statement has nothing to do with this.

For the stack overflow, probably the best span to use would be that of the call site.

**Review by RalfJung — COMMENTED** [review]
(wow GitHub is really making my life as hard as possible today. now it started a review...)

**Comment by RalfJung** [comment]
All right, I got it. I moved the stack check to *before* the new frame is added, so the span is at the call site.

@oli-obk probably better if you check this again since I did a bit more than just fixing those typos.

File: src/librustc_mir/const_eval/machine.rs [hunk 4]
```diff
@@ -301,12 +301,16 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
...
-    fn after_stack_push(ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
-        // Enforce stack size limit.
-        if !ecx.tcx.sess.recursion_limit().value_within_limit(ecx.stack().len()) {
+    #[inline(always)]
+    fn init_frame_extra(
+        ecx: &mut InterpCx<'mir, 'tcx, Self>,
+        frame: Frame<'mir, 'tcx>,
+    ) -> InterpResult<'tcx, Frame<'mir, 'tcx>> {
+        // Enforce stack size limit. Add 1 because this is run before the new frame is pushed.
+        if !ecx.tcx.sess.recursion_limit().value_within_limit(ecx.stack().len()+1) {
```
Comment (oli-obk) [inline src/librustc_mir/const_eval/machine.rs:310]: rustfmt :laughing: 
```suggestion
        if !ecx.tcx.sess.recursion_limit().value_within_limit(ecx.stack().len() + 1) {
```

Comment (RalfJung) [reply src/librustc_mir/const_eval/machine.rs:310]: argh...

File: src/librustc_mir/interpret/eval_context.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:698]: yea, the body's span is good here, though it's actually the call site's fault. But lets leave such refactorings (using the call site span) to the future, I remember trying something similar a while back and it just complicated the span scheme even more

Comment (RalfJung) [reply src/librustc_mir/interpret/eval_context.rs:698]: Hm, this wouldn't even be that hard I think... we can just call `cur_span` before pushing the new frame, then we should get the call site.

Comment (RalfJung) [reply src/librustc_mir/interpret/eval_context.rs:698]: It's just not clear to me that we always want the call site. The callee could also be responsible for the problem -- in fact I'd usually think they are.

Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs:698]: at runtime the decision whose job it is to create the stack frame is an ABI decision, but that doesn't really make sense for CTFE. But I also think your alluding to a different thing? If an required_const will fail regardless of the concrete generic parameters, then that's indeed the function's fault and not the caller's

Comment (RalfJung) [reply src/librustc_mir/interpret/eval_context.rs:698]: > If an required_const will fail regardless of the concrete generic parameters, then that's indeed the function's fault and not the caller's

Yes, that is one possible scenario I had in mind where blaming the caller seems wrong.

**Comment by RalfJung** [comment]
```
error: embedded-book maintainer @ryankurte is not assignable in the rust-lang/rust repo
error: embedded-book maintainer @thejpster is not assignable in the rust-lang/rust repo
error: embedded-book maintainer @korken89 is not assignable in the rust-lang/rust repo
```
Seems spurious to me... I hope.^^
@bors retry

**Comment by RalfJung** [comment]
@rust-lang/infra
second time in a row that I got
```
error: embedded-book maintainer @thejpster is not assignable in the rust-lang/rust repo
error: embedded-book maintainer @korken89 is not assignable in the rust-lang/rust repo
error: embedded-book maintainer @ryankurte is not assignable in the rust-lang/rust repo
```
Any idea what is happening?
@bors retry

**Comment by pietroalbini** [comment]
Fixing CI in https://github.com/rust-lang/rust/pull/75450.

**Pull request merged**